osquery/tables/system/windows/wmi.cpp

```diff
@@ -0,0 +1,306 @@
+  if (hr != S_OK) {
+    ret = "";
+    return Status(-1, "Error retrieving data from WMI query.");
+  }
+  if (value.vt != VT_BSTR) {
+    ret = "";
+    return Status(-1, "Invalid data type returned.");
+  }
+  ret = BSTR_to_string(value.bstrVal);
+  VariantClear(&value);
+  return Status(0);
+}
+
+Status WmiResultItem::GetVectorOfStrings(const std::string& name,
+                                         std::vector<std::string>& ret) const {
+  std::wstring property_name = string_to_wstring(name);
+  VARIANT value;
+  HRESULT hr = result_->Get(property_name.c_str(), 0, &value, nullptr, nullptr);
+  if (hr != S_OK) {
+    return Status(-1, "Error retrieving data from WMI query.");
+  }
+  if (value.vt != (VT_BSTR | VT_ARRAY)) {
+    return Status(-1, "Invalid data type returned.");
+  }
+  long lbound, ubound;
+  SafeArrayGetLBound(value.parray, 1, &lbound);
+  SafeArrayGetUBound(value.parray, 1, &ubound);
+  long count = ubound - lbound + 1;
+
+  BSTR* pData = nullptr;
+  SafeArrayAccessData(value.parray, (void**)&pData);
+  ret.reserve(count);
+  for (long i = 0; i < count; i++) {
+    ret.push_back(BSTR_to_string(pData[i]));
+  }
+  SafeArrayUnaccessData(value.parray);
+  VariantClear(&value);
+  return Status(0);
+}
+
+WmiRequest::WmiRequest(const std::string& query) {
+  std::wstring wql = string_to_wstring(query);
+
+  HRESULT hr = E_FAIL;
+
+  hr = ::CoInitializeEx(0, COINIT_MULTITHREADED);
+  hr = ::CoInitializeSecurity(nullptr,
+                              -1,
+                              nullptr,
+                              nullptr,
+                              RPC_C_AUTHN_LEVEL_DEFAULT,
+                              RPC_C_IMP_LEVEL_IMPERSONATE,
+                              nullptr,
+                              EOAC_NONE,
+                              nullptr);
+  hr = ::CoCreateInstance(CLSID_WbemLocator,
+                          0,
+                          CLSCTX_INPROC_SERVER,
+                          IID_IWbemLocator,
+                          (LPVOID*)&locator_);
+  if (hr != S_OK) {
+    locator_ = nullptr;
+    return;
+  }
+
+  hr = locator_->ConnectServer(L"ROOT\\CIMV2",
+                               nullptr,
+                               nullptr,
+                               nullptr,
+                               0,
+                               nullptr,
+                               nullptr,
+                               &services_);
+  if (hr != S_OK) {
+    services_ = nullptr;
+    return;
+  }
+
+  hr = services_->ExecQuery(
+      L"WQL", (BSTR)wql.c_str(), WBEM_FLAG_FORWARD_ONLY, nullptr, &enum_);
+  if (hr != S_OK) {
+    enum_ = nullptr;
+    return;
+  }
+
+  hr = WBEM_S_NO_ERROR;
+  while (hr == WBEM_S_NO_ERROR) {
+    IWbemClassObject* result = nullptr;
+    ULONG result_count = 0;
+
+    hr = enum_->Next(WBEM_INFINITE, 1, &result, &result_count);
+    if (SUCCEEDED(hr) && result_count > 0) {
+      results_.push_back(WmiResultItem(result));
```

Comment:
There is resource manipulation within the ctor here. I wonder if the input type needs memory handling before it goes out of scope?